### PR TITLE
feat: add blockstore that is the union of several CAR stores

### PR DESCRIPTION
**Pull request by lemmih**
## Summary of changes

Changes introduced in this pull request:

- add `ManyCar` blockstore which takes the union of several CAR-files.
- use a shared z-frame cache, limited by byte size rather than pages.
- use `ManyCar` in as many commands as possible.

## Reference issue to close (if applicable)

Roadmap: #3222

## Benchmarks

Linear scan through a mainnet snapshot to find the genesis block:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `Genesis Scan: 0-MiB` | 7.829 ± 0.120 | 7.629 | 7.975 | 3.38 ± 0.06 |
| `Genesis Scan: 32-MiB` | 2.314 ± 0.019 | 2.287 | 2.354 | 1.00 |
| `Genesis Scan: 64-MiB` | 2.342 ± 0.031 | 2.316 | 2.401 | 1.01 ± 0.02 |
| `Genesis Scan: 128-MiB` | 2.401 ± 0.037 | 2.357 | 2.458 | 1.04 ± 0.02 |
| `Genesis Scan: 256-MiB` | 2.436 ± 0.019 | 2.400 | 2.473 | 1.05 ± 0.01 |
| `Genesis Scan: 512-MiB` | 2.522 ± 0.036 | 2.489 | 2.573 | 1.09 ± 0.02 |
| `Genesis Scan: 1024-MiB` | 2.542 ± 0.024 | 2.503 | 2.573 | 1.10 ± 0.01 |
| `Genesis Scan: 2048-MiB` | 2.523 ± 0.036 | 2.488 | 2.602 | 1.09 ± 0.02 |

Caching a single frame is sufficient when doing a linear scan. It's surprising that large caches hurt performance, though. Not sure why that is.

Executing 100 tipsets:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `Execute 100 tipsets: 0-MiB` | 212.465 ± 1.082 | 211.468 | 213.615 | 3.33 ± 0.02 |
| `Execute 100 tipsets: 32-MiB` | 71.316 ± 1.105 | 70.204 | 72.415 | 1.12 ± 0.02 |
| `Execute 100 tipsets: 64-MiB` | 68.967 ± 0.652 | 68.378 | 69.667 | 1.08 ± 0.01 |
| `Execute 100 tipsets: 128-MiB` | 66.412 ± 0.174 | 66.265 | 66.604 | 1.04 ± 0.01 |
| `Execute 100 tipsets: 256-MiB` | 66.021 ± 0.080 | 65.947 | 66.106 | 1.04 ± 0.00 |
| `Execute 100 tipsets: 512-MiB` | 63.754 ± 0.270 | 63.504 | 64.039 | 1.00 |
| `Execute 100 tipsets: 1024-MiB` | 63.764 ± 0.357 | 63.391 | 64.102 | 1.00 ± 0.01 |
| `Execute 100 tipsets: 2048-MiB` | 64.399 ± 0.790 | 63.865 | 65.306 | 1.01 ± 0.01 |

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/db/car/any.rs [hunk 1]
```diff
@@ -1,25 +1,43 @@
...
-impl<ReaderT: Read + Seek> AnyCar<ReaderT> {
-    pub fn new(mk_reader: impl Fn() -> Result<ReaderT> + Clone + 'static) -> Result<Self> {
-        if let Ok(forest_car) = super::ForestCar::new(mk_reader.clone()) {
+impl<ReaderT: super::CarReader> AnyCar<ReaderT> {
+    /// Open an archive. May be formatted as `.car`, `.car.zst` or
+    /// `.forest.car.zst`. This call may block for an indeterminate amount of
+    /// time while data is decoded and indexed.
+    pub fn new(mk_reader: impl super::forest::ReaderGen<ReaderT>) -> Result<Self> {
+        let plain_reader = mk_reader();
+        let zstd_reader = mk_reader();
```
Comment (hanabi1224) [inline src/db/car/any.rs:32]: `zstd_reader` does not seem to be used in plain car code branch, how about inlining these

File: src/db/car/many.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        ManyCar {
+            shared_cache: Arc::new(Mutex::new(ZstdFrameCache::default())),
+            read_only: Vec::new(),
+            writer: MemoryDB::default(),
+        }
+    }
+}
+
+impl<WriterT> ManyCar<WriterT> {
+    pub fn read_only<ReaderT: super::CarReader>(&mut self, any_car: AnyCar<ReaderT>) {
```
Comment (hanabi1224) [inline src/db/car/many.rs:37]: nit: `push_read_only` might be more descriptive

File: src/db/car/many.rs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+use parking_lot::Mutex;
+use std::{io, path::PathBuf, sync::Arc};
+
+pub struct ManyCar<WriterT = MemoryDB> {
+    shared_cache: Arc<Mutex<ZstdFrameCache>>,
+    read_only: Vec<AnyCar<Box<dyn super::CarReader>>>,
+    writer: WriterT,
+}
+
+impl ManyCar {
```
Comment (hanabi1224) [inline src/db/car/many.rs:26]: nit: [state migration tests](https://github.com/ChainSafe/forest/blob/main/src/state_migration/tests/mod.rs#L91) could use `ManyCar`. I can also do that in a separate PR

**Review by hanabi1224 — APPROVED** [review]

File: src/db/car/any.rs (hunk 1 above)
Comment (lemmih) [reply src/db/car/any.rs:32]: Cannot be moved. `mk_reader` is being consumed one line below.

**Review by LesnyRumcajs — COMMENTED** [review]
Adding subcommand tests, at least for the happy path, would be great.

**Comment by lemmih** [comment]
> Adding subcommand tests, at least for the happy path, would be great.

I can add smoke tests once the [`--diff`](https://github.com/ChainSafe/forest/pull/3284) PR is merged.

File: src/db/car/many.rs (hunk 3 above)
Comment (hanabi1224) [reply src/db/car/many.rs:26]: nvm, bundle car is no longer needed once https://github.com/ChainSafe/forest/pull/3276 is merged

**Comment by LesnyRumcajs** [comment]
> > Adding subcommand tests, at least for the happy path, would be great.
> 
> I can add smoke tests once the [`--diff`](https://github.com/ChainSafe/forest/pull/3284) PR is merged.

What's preventing us from doing a test without the `--diff`?

**Comment by lemmih** [comment]
> > > Adding subcommand tests, at least for the happy path, would be great.
> > 
> > 
> > I can add smoke tests once the [`--diff`](https://github.com/ChainSafe/forest/pull/3284) PR is merged.
> 
> What's preventing us from doing a test without the `--diff`?

I don't have two CAR files to merge. :)

The primary use case for `ManyCar` is to merge a base snapshot and a set of diffs into a single, coherent block store. Without the `--diff` command, I'm not sure how to test it (other than the unit tests).

**Comment by lemmih** [comment]
> Adding subcommand tests, at least for the happy path, would be great.

Done. Happy path tested.

**Comment by lemmih** [comment]
> @lemmih Sadly, the happy path is not so happy. https://github.com/ChainSafe/forest/actions/runs/5712759803/job/15477232048?pr=3288

See!? This is why we should never add any tests! :P

File: src/db/car/forest.rs [hunk 4]
```diff
@@ -126,57 +176,76 @@ where
...
+        let indexed = hoist_error(&mut *indexed_guard)?;
+
+        for position in indexed.lookup(*k)?.into_iter() {
+            let reader = indexed.get_mut();
+            let cache_query = self.frame_cache.lock().get(position, self.cache_key, *k);
+            match cache_query {
+                // Frame cache hit, found value.
+                Some(Some(val)) => return Ok(Some(val)),
+                // Frame cache hit, no value. This only happens when hashes collide
+                Some(None) => {}
```
Comment (LesnyRumcajs) [inline src/db/car/forest.rs:194]: Do we expect hash collisions? If not, should we panic here?

Comment (lemmih) [reply src/db/car/forest.rs:194]: Hash collisions may occur, and it costs us little to handle them correctly. If two CIDs have the same hash, the index returns two possible positions. We check the second position if we don't find the correct CID at the first position.

**Review by LesnyRumcajs — APPROVED** [review]

**Pull request merged**